components/tikv_alloc/src/jemalloc.rs

```diff
@@ -21,18 +27,125 @@ lazy_static! {
         Mutex::new(HashMap::new());
 }
 
+thread_local! {
+    static MEMORY_STAT_ACCESSOR_STILL_ALIVE: AccessorStillAlive = {
```

Comment:
How about implementing `Drop` for `MemoryStatsAccessor`, and call a function like `take` to mark consumed in `remove_thread_memory_accessor` before dropping the `MemoryStatsAccessor`. So you check whether it is consumed in drop without another map

Reply:
As @tabokie described, I think we can directly store `MemoryStatsAccessor` in TLS and then we can fully get rid of the `THREAD_MEMORY_MAP`. As a side effect the `remove_memory_accessor` can be trivially no-op.

Reply:
you mean https://docs.rs/thread_local/latest/thread_local/struct.ThreadLocal.html ? But seems it doesn't drop the per-thread objects at thread exit

Reply:
Well, so I think we still need someway to hook at the thread ends if we want to keep safety. And the `ThreadLocal` may not be usable for this scenario because we cannot remove the accessor from the map when a thread exits(which may leak memory if there are too many new threads spawn, e.g. blocking threads of tokio.) cc @tabokie

Reply:
then test would detect this

Reply:
> And the ThreadLocal may not be usable for this scenario

I agree.

> I think the problem is how can we make the MemoryStatsAccessor being removed from the trace list even someone forget to call remove_memory_stat_accessor

You should limit the use of this inside tikv_util::spawn_wrapper. And disallow any use of it outside tikv_util with clippy rule.

Reply:
@Connor1996 I have noticed that when a test case finished, the variables created by `lazy_static!` won't be dropped. So I guess the tests cannot detect the leakage. 😢

Reply:
why isn't it dropped?

Reply:
I'm not sure, perhaps because variables created by `lazy_static!` (the `MEMORY_MAP`) lives for `'static` (which equals to the lifetime of the whole program?)